configsuite/schema.py

```diff
@@ -40,11 +40,14 @@ def _check_allownone_type(schema_level):
     return True
 
 
-@configsuite.validator_msg("Non required types must allow None")
+@configsuite.validator_msg(
+    "Non required types must allow None or have non-None default"
+)
 def _check_allownone_required(schema_level):
     if isinstance(schema_level[MK.Type], types.BasicType):
         allow_none = schema_level.get(MK.AllowNone, False)
-        if not allow_none:
+        has_non_none_default = schema_level.get(MK.Default) is not None
+        if not allow_none and not has_non_none_default:
             return schema_level.get(MK.Required, True)
     return True
```

Comment:
Is the checking of type necessary at all? This is done in the `_check_allownone_type`. I suppose the `ElementValidators` are all checked, and not just break on first `False`?

Other than that, how about rewriting it a bit? This way you can pretty much see how the validator_msg aligns with the code.
```
if not schema_level.get(MK.Required, True):
    allow_none = schema_level.get(MK.AllowNone, False)
    has_non_none_default = schema_level.get(MK.Default) is not None
    return allow_none or has_non_none_default
return True
```

Reply:
The type check prevents the error of having an invalid combination of required, allowing none and default values for a container, as this will never be relevant. The only real error is that allowing none is not supported for containers and hence I prefer if that is the only error reported.

Reply:
How would you feel about:

```python
is_required = schema_level.get(MK.Required, True)
is_basic_type = isinstance(schema_level[MK.Type]types.BasicType)
if is_required or not is_basic_type:
    return True

allow_none = schema_level.get(MK.AllowNone, False)
has_non_none_default = schema_level.get(MK.Default) is not None
return allow_none or has_non_none_default
```

Reply:
> The only real error is that allowing none is not supported for containers and hence I prefer if that is the only error reported.

I agree! The error messages should definitively be as relevant as possible

Implementation looks good!

Reply:
Just changed the implementation ;)